modules/dlm/src/main/java/org/elasticsearch/dlm/ModifyDataLifecycleService.java

```diff
@@ -0,0 +1,151 @@
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateAckListener;
+import org.elasticsearch.cluster.ClusterStateTaskExecutor;
+import org.elasticsearch.cluster.ClusterStateTaskListener;
+import org.elasticsearch.cluster.SimpleBatchedAckListenerTaskExecutor;
+import org.elasticsearch.cluster.metadata.DataLifecycle;
+import org.elasticsearch.cluster.metadata.DataStream;
+import org.elasticsearch.cluster.metadata.IndexAbstraction;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.cluster.service.MasterServiceTaskQueue;
+import org.elasticsearch.common.Priority;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.core.Tuple;
+
+import java.util.List;
+
+/**
+ * This service modifies the data lifecycle of an existing data stream. It creates and batches the cluster state update tasks
+ * that will change the data stream metadata in the cluster state.
+ */
+public class ModifyDataLifecycleService {
+
+    private final MasterServiceTaskQueue<ModifyLifecycleTask> taskQueue;
+
+    @Inject
+    public ModifyDataLifecycleService(ClusterService clusterService) {
+        ClusterStateTaskExecutor<ModifyLifecycleTask> executor = new SimpleBatchedAckListenerTaskExecutor<>() {
+
+            @Override
+            public Tuple<ClusterState, ClusterStateAckListener> executeTask(
+                ModifyLifecycleTask modifyLifecycleTask,
+                ClusterState clusterState
+            ) {
+                return new Tuple<>(
+                    modifyLifecycle(clusterState, modifyLifecycleTask.dataStreamNames(), modifyLifecycleTask.dataLifecycle()),
+                    modifyLifecycleTask
+                );
+            }
+        };
+        this.taskQueue = clusterService.createTaskQueue("modify-lifecycle", Priority.LOW, executor);
+    }
+
+    /**
+     * Submits the task to set the lifecycle to the requested data streams.
+     */
+    public void setLifecycle(
+        final List<String> dataStreamNames,
+        DataLifecycle lifecycle,
+        TimeValue ackTimeout,
+        final ActionListener<AcknowledgedResponse> listener
+    ) {
+        // TODO use request.masterNodeTimeout() as timeout?
+        taskQueue.submitTask("set-lifecycle", new ModifyLifecycleTask(dataStreamNames, lifecycle, ackTimeout, listener), null);
+    }
+
+    /**
+     * Submits the task to remove the lifecycle from the requested data streams.
+     */
+    public void removeLifecycle(List<String> dataStreamNames, TimeValue ackTimeout, ActionListener<AcknowledgedResponse> listener) {
+        // TODO use request.masterNodeTimeout() as timeout?
```

Comment:
++ the master timeout could be a parameter for this method (same for others)

Reply:
I will add it